src/proxy/http/balance.rs

```diff
@@ -118,19 +152,426 @@ where
+
+pub mod fallback {
+    use bytes::Buf;
+    use futures::{future, Async, Future, Poll};
+    use hyper::body::Payload;
+
+    use http;
+    use proxy::{
+        buffer,
+        http::router,
+        resolve::{EndpointStatus, HasEndpointStatus},
+    };
+    use svc;
+
+    use super::Error;
+
+    use std::marker::PhantomData;
+
+    extern crate linkerd2_router as rt;
+
+    #[derive(Debug, Clone)]
+    pub struct Layer<Rec, Bal, A> {
+        recognize: Rec,
+        max_in_flight: usize,
+        balance_layer: Bal,
+        _marker: PhantomData<fn(A)>,
+    }
+
+    #[derive(Debug)]
+    pub struct Stack<R, Bal, A> {
+        fallback: R,
+        balance: Bal,
+        _marker: PhantomData<fn(A)>,
+    }
+
+    #[derive(Debug)]
+    pub struct MakeSvc<R, Bal, A> {
+        fallback: MakeCurried<R, router::Config>,
+        balance: Bal,
+        _marker: PhantomData<fn(A)>,
+    }
+
+    #[derive(Debug)]
+    pub struct MakeFuture<R, F, A>
+    where
+        F: Future,
+    {
+        fallback: Option<MakeCurried<R, router::Config>>,
+        making: F,
+        _marker: PhantomData<fn(A)>,
+    }
+
+    pub struct Service<F, Bal, A>
+    where
+        F: rt::Make<router::Config>,
+    {
+        mk_fallback: MakeCurried<F, router::Config>,
+        fallback: Option<F::Value>,
+        balance: Bal,
+        status: EndpointStatus,
+        _marker: PhantomData<fn(A)>,
+    }
+
+    #[derive(Debug, Clone)]
+    struct MakeCurried<Mk, T> {
+        mk: Mk,
+        target: T,
+    }
+
+    #[derive(Clone, Debug)]
+    pub enum Body<A, B> {
+        A(A),
+        B(B),
+    }
+
+    pub fn layer<Rec, A, B, D>(
+        balance_layer: super::Layer<A, B, D>,
+        max_in_flight: usize,
+        recognize: Rec,
+    ) -> Layer<Rec, super::Layer<A, B, D>, A>
+    where
+        Rec: router::Recognize<http::Request<A>> + Clone,
+    {
+        Layer {
+            recognize,
+            max_in_flight,
+            balance_layer,
+            _marker: PhantomData,
+        }
+    }
+
+    impl<Rec, Bal, A, M> svc::Layer<M> for Layer<Rec, Bal, A>
+    where
+        Rec: router::Recognize<http::Request<A>> + Clone + Send + Sync + 'static,
+        router::Layer<http::Request<A>, Rec>:
+            svc::Layer<<buffer::Layer<http::Request<A>> as svc::stack::Layer<M>>::Service>,
+        buffer::Layer<http::Request<A>>: svc::Layer<M>,
+        Bal: svc::Layer<M>,
+        M: Clone,
+    {
+        type Service = Stack<
+            <router::Layer<http::Request<A>, Rec> as svc::Layer<
+                <buffer::Layer<http::Request<A>> as svc::stack::Layer<M>>::Service,
+            >>::Service,
+            Bal::Service,
+            A,
+        >;
+
+        fn layer(&self, inner: M) -> Self::Service {
+            let balance = self.balance_layer.layer(inner.clone());
+            let inner = buffer::layer(self.max_in_flight).layer(inner);
+            let fallback = router::layer(self.recognize.clone()).layer(inner);
```

Comment:
Initially, I tried to use `ServiceBuilder` here but couldn't figure out the bounds to make that work.